blog/posts/2020-07-09-miragejs-graphql-tools.md

```diff
@@ -0,0 +1,198 @@
+```
+
+You can then add that do the configuration for Mirage as `models: modelMaps` and you'll get that automatically registered as you add to your schema. This does get more complicated though, when we start to add associations in our objects and need Mirage to see that as a relationship for queries that load all that data. Ideally, the graph can work for a query like so on the UI:
+
+```graphql
+query ListAllMovies {
+  movies {
+    actors {
+      name      
+    }
+    distributor {
+      name
+    }
+    title
+  }
+}
+```
+
+We first want to identify all the model names (variable `modelNames`). Also, you want to reduce the fields you're checking against to only fields that are confirmed to be other object types (variable `modelsReducedFields`). 
+
+```js
+  const modelNames = nodeTypes.map(type => type.model);
+  
+  const modelsReducedFields = nodeTypes.map(node => {
+    const nodeFields = node.fields || [];
+    const fields = nodeFields.reduce((acc, field) => {
+      const { type } = field;
+
+      const isNestedType = node => !node.name && node.type;
+
+      if (isNestedType(type)) {
+        const rootField = _getRootType(field);
+        const isListType = field.type.type.kind === "ListType";
+        const model = rootField.name.value;
+        if (modelNames.includes(model)) {
+          acc.push({
+            name: field.name.value,
+            model,
+            isListType
+          });
+        }
+        return acc;
+      }
+
+      return acc;
+    }, []);
+  return { ...node, fields };
+});
+```
+
+Now, to walk through what we're doing here with `modelsReducedFields()` is to take each node and reduce the fields down to other models and if it's a belongs-to or has-many kind of association. You might have notices the call to `_getRootType()`, which is just a recursive function to go through nested objects in the AST and get the deepest node's name. I'm showing it independently in the following:
+
+```js
+const _getRootType = field => (field.type ? _getRootType(field.type) : field);
+```
+
+We can now use this improved array for the `modelMaps` value to get models that have the associations automatically created.
+
+```js
+const modelMaps = modelsReducedFields.reduce((modelAccumulator, node) => {
+  modelAccumulator[node.model] = Model;
+  if (node.fields.length) {
+    // any remaining field we know has a model as well
+    const fields = node.fields.reduce((fieldsAcc, field) => {
+      
+      fieldsAcc[field.name] = field.isListType
+        ?  hasMany(field.model)
+        : belongsTo(field.model);
+      return fieldsAcc;
+    }, {});
+
+    modelAccumulator[node.model] = Model.extend(fields);
+  }
+
+  return modelAccumulator;
+}, {});
+```
+
+While this seems like a lot up front for such a small schema. If your schema is larger, or you expect your types to scale, yet still want Mirage for simulation or testing coverage, then it's really helpful to get this up front without maintaining those changes over time. 
```

Comment:
```suggestion
While this may seem like a lot up front for such a small schema, if your schema is larger, or you expect your types to scale, yet still want to use Mirage for simulation or testing coverage, then it's really helpful to get this up front without maintaining those changes over time.
```